index.d.ts

```diff
@@ -0,0 +1,81 @@
+export interface Action {
+  type: string;
+}
+
+
+/* reducers */
+
+export type Reducer<S> = <A extends Action>(state: S, action: A) => S;
+
+export function combineReducers<S>(reducers: {[key: string]: Reducer<any>}): Reducer<S>;
+
+
+/* store */
+
+export interface Dispatch {
+  <A>(action: A): A;
+  <A, B>(action: A): B;
+}
+
+export interface Store<S> {
+  dispatch: Dispatch;
+  getState(): S;
+  subscribe(listener: () => void): () => void;
+  replaceReducer(reducer: Reducer<S>): void;
+}
+
+export interface StoreCreator {
+  <S>(reducer: Reducer<S>): Store<S>;
+  <S>(reducer: Reducer<S>, initialState: S): Store<S>;
+
+  <S>(reducer: Reducer<S>, enhancer: StoreEnhancer): Store<S>;
+  <S>(reducer: Reducer<S>, initialState: S, enhancer: StoreEnhancer): Store<S>;
+}
+
+export type StoreEnhancer = (next: StoreCreator) => StoreCreator;
+
+export const createStore: StoreCreator;
+
+
+/* middleware */
+
+export interface MiddlewareAPI<S> {
+  dispatch: Dispatch;
+  getState(): S;
+}
+
+export interface Middleware {
+  <S>(api: MiddlewareAPI<S>): (next: Dispatch) => <A, B>(action: A) => B;
+}
+
+export function applyMiddleware(...middlewares: Middleware[]): StoreEnhancer;
+
+
+/* action creators */
+
+export interface ActionCreator {
+  <O>(...args: any[]): O;
+}
+
+export function bindActionCreators<
+  T extends ActionCreator|{[key: string]: ActionCreator}
+>(actionCreators: T, dispatch: Dispatch): T;
+
+
```

Comment:
> ```ts
> <A extends ActionCreator<any>>(actionCreator: A, dispatch: Dispatch): A;
> ```
> is not typesafe as it enforces any.

I'm not sure I understand. This signature allows doing
```ts
bindActionCreators<ActionCreator<MyAction>>(...)
```
and even
```ts
bindActionCreators<(text: string) => MyAction>(...)
```
So type parameter is not enforced.

In contrast, if we used this signature:
```ts
<A>(actionCreator: ActionCreator<A>, dispatch: Dispatch): ActionCreator<A>;
```
to bind action creator of type `(text: string) => MyAction`, then `bindActionCreators` would return value of type `(...args: any[]) => MyAction`, effectively erasing constraints on its arguments.

Also, I found that we don't cover cases when `bindActionCreators` argument and return value have different types, e.g. if action creator returns thunk, then bound action creator will return what thunk returned.